README.md

# QUANTAXIS 量化金融策略框架

------------------



点击此处下载 QUANTAXIS2019 介绍性pdf [quantaxis 2019](http://whl.yutiansut.com/QUANTAXIS%202019.pdf)


Quantitative Financial FrameWork

从数据爬取-清洗存储-分析回测-可视化-交易复盘的本地一站式解决方案

![QUANTAXIS_LOGO_LAST_small.jpg](http://pic.yutiansut.com/Fn0TPEcwu_uhraf58_93Ul5yfvAz)

![QUANTAXIS 2019.png](http://pic.yutiansut.com/FnRlMW2LQpFBrsdRv7E_uJ9RvzHt)

![pypidownloads](https://img.shields.io/pypi/dm/quantaxis.svg)
![pypidownloads](https://img.shields.io/pypi/dw/quantaxis.svg)
[![Backers on Open Collective](https://opencollective.com/QUANTAXIS/backers/badge.svg)](#backers) [![Sponsors on Open Collective](https://opencollective.com/QUANTAXIS/sponsors/badge.svg)](#sponsors) 

[![Github workers](https://img.shields.io/github/watchers/quantaxis/quantaxis.svg?style=social&label=Watchers&)](https://github.com/quantaxis/quantaxis/watchers)
[![GitHub stars](https://img.shields.io/github/stars/quantaxis/quantaxis.svg?style=social&label=Star&)](https://github.com/quantaxis/quantaxis/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/quantaxis/quantaxis.svg?style=social&label=Fork&)](https://github.com/quantaxis/quantaxis/fork)

[点击右上角Star和Watch来跟踪项目进展! 点击Fork来创建属于你的QUANTAXIS!]


## 祝贺 QUANTAXIS 入选 [GITEE 最有价值开源项目 GVP](https://gitee.com/gvp)
![gvp](http://pic.yutiansut.com/gvp.jpg)



![main_1](http://pic.yutiansut.com/Main_1.gif)
![presentbyyutiansut](http://pic.yutiansut.com/yutiansut-logo.png)



![version](https://img.shields.io/pypi/v/quantaxis.svg)
![build](https://travis-ci.org/QUANTAXIS/QUANTAXIS.svg?branch=master)
[![Codefresh build status]( https://g.codefresh.io/api/badges/pipeline/quantaxis/QUANTAXIS%2FQUANTAXIS%2FQUANTAXIS?branch=master&key=eyJhbGciOiJIUzI1NiJ9.NWEzMGMwNzE4NTk5YzAwMDAxNWEzOGQ3.INF0oVNGLK79Ql3f10Vs9_CJfwsfq1FYjCbbOKJzBe8&type=cf-1)]( https://g.codefresh.io/pipelines/QUANTAXIS/builds?repoOwner=QUANTAXIS&repoName=QUANTAXIS&serviceName=QUANTAXIS%2FQUANTAXIS&filter=trigger:build~Build;branch:master;pipeline:5c3ade3fc67fe4e3328c9da1~QUANTAXIS)
[![BCH compliance](https://bettercodehub.com/edge/badge/QUANTAXIS/QUANTAXIS?branch=master)](https://bettercodehub.com/)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/d8504e4af33747bb8117579212425af9)](https://www.codacy.com/app/yutiansut/QUANTAXIS?utm_source=github.com&utm_medium=referral&utm_content=yutiansut/QUANTAXIS&utm_campaign=badger)
[![StackShare](https://img.shields.io/badge/tech-stack-0690fa.svg?style=flat)](https://stackshare.io/yutiansut/quantaxis)
![QAS](https://img.shields.io/badge/QAS-%200.0.8-brown.svg)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/QUANTAXIS/QUANTAXIS/master)
![python](https://img.shields.io/badge/python-%203.7/3.6-darkgrey.svg)
![Npm](https://img.shields.io/badge/Npm-%200.4.0-yellow.svg)
![author](https://img.shields.io/badge/Powered%20by-%20%20yutiansut-red.svg)
![license](https://img.shields.io/badge/License-%20MIT-brightgreen.svg)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FQUANTAXIS%2FQUANTAXIS.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2FQUANTAXIS%2FQUANTAXIS?ref=badge_shield)
[![Waffle.io - Issues in progress](https://badge.waffle.io/yutiansut/QUANTAXIS.png?label=in%20progress&title=In%20Progress)](http://waffle.io/yutiansut/QUANTAXIS)
[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)
[![](http://rescdn.qqmail.com/zh_CN/htmledition/images/function/qm_open/ico_mailme_22.png)](http://mail.qq.com/cgi-bin/qm_share?t=qm_mailme&email=VSwgITw0OyYgIRUjPCV7JCR7Njo4)

## Contributors

This project exists thanks to all the people who contribute. [[Contribute](https://github.com/QUANTAXIS/QUANTAXIS/graphs/contributors)].
<a href="https://github.com/QUANTAXIS/QUANTAXIS/graphs/contributors"><img src="https://opencollective.com/QUANTAXIS/contributors.svg?width=890&button=false" /></a>


> 欢迎加群讨论: 563280067 [群链接](https://jq.qq.com/?_wv=1027&k=4CEKGzn) 

> QUANTAXIS 开发群: 773602202 (如果想要贡献代码 请加这个群 需要备注你的GITHUB ID)

> 欢迎关注公众号: ![公众号](http://data.yutiansut.com/qrcode_for_gh_bbb47e0550f7_258.jpg)

QAPRO公共号免费提供了下单推送接口, 关注公共号回复trade即可使用

> 许多问题 可以在 [GITHUB ISSUE](https://github.com/QUANTAXIS/QUANTAXIS/issues)中找到, 你可以提出新的issue

> QUANTAXIS 内测版论坛 [QUANTAXISCLUB上线](http://www.yutiansut.com:3000)

http://www.yutiansut.com:3000

凡通过论坛进行提问的 均有最高的回复优先级

> 全新文档界面 [QUANTAXISDocs](http://doc.yutiansut.com)

http://doc.yutiansut.com



QUANTAXIS量化金融策略框架,是一个面向中小型策略团队的量化分析解决方案. 我们通过高度解耦的模块化以及标准化协议,可以快速的实现面向场景的定制化解决方案.QUANTAXIS是一个渐进式的开放式框架,你可以根据自己的需要,引入自己的数据,分析方案,可视化过程等,也可以通过RESTful接口,快速实现多人局域网/广域网内的协作.

==========================================


## 关联项目:

技术栈: python/nodejs/vue/mongodb/rabbitmq/c++

### 核心工具链(生产环境在用)

![QQ图片20191029223640.png](http://pic.yutiansut.com/FuVrzcbWJUBNrj4Wa0zlRl-YlBY_)

#### 已开源

> 数据存储/数据分析/回测

- [QUANTAXIS](https://github.com/QUANTAXIS/QUANTAXIS) QUANTAXIS的核心部分

> WEB相关, http/websocket/开放数据接口

- [QUANTAXIS_WEBSERVER](https://github.com/QUANTAXIS/QUANTAXIS_WEBSERVER) 基于tornado的web api/ websocket

> 分布式相关, 任务异步执行, 跨进程分布式消息订阅分发

- [QUANTAXIS_RUN](https://github.com/QUANTAXIS/quantaxis_run) 基于rabbitmq/celery的分布式任务部署
- [QUANTAXIS_PUBSUB](https://github.com/QUANTAXIS/QAPUBSUB) 基于RABBITMQ的消息分发订阅

> 接口相关: 交易账户/ 期货接口封装/ Trader实例
- [QUANTAXIS OTGBROKER](https://github.com/QUANTAXIS/QAOTGBROKER) 基于OPEN_TRADE_GATEWAY的接口封装
- [QUANTAXIS CTPBEEBROKER](https://github.com/QUANTAXIS/QACTPBeeBroker) 基于CTPBee的接口封装
- [QUANTAXIS_ATBROKER](https://github.com/QUANTAXIS/QA_AtBroker) 基于海风at的接口封装
- [QUANTAXIS TRADER](https://github.com/yutiansut/QATrader) 一个开源的websocket版本的期货交易实例

> 策略相关 
- [QASTRATEGY101](https://github.com/yutiansut/QAStrategy101) 101个基础策略[逐步更新中...]

> 行情相关: 主推行情实现/ 基于OU过程的模拟行情
- [QUNATAXIS MARKETCOLLECTOR](https://github.com/yutiansut/QUANTAXIS_RealtimeCollector) 全市场订阅分发的行情推送
- [QUANTAXIS_RandomPrice](https://github.com/yutiansut/QUANTAXIS_RandomPrice) 基于OU过程的随机行情模拟

> 账户协议

- [QIFI](https://github.com/QUANTAXIS/QIFI) 一个基于快期DIFF协议的QA实时账户协议
- [QIFIAccount](https://github.com/yutiansut/qifiaccount) 一个基于QIFI协议的多市场兼容的 实时账户实现
- [QAStrategy](https://github.com/yutiansut/qastrategy) 一个完整的 支持 模拟/回测/实盘一键切换的策略基类

#### 未开源

> 实时交易解决方案/ 无人值守/状态汇报/实时账户评估/多账户/策略账户拆分/事件流风控/PB系统/CEP引擎/多系统终端

- [QUANTAXIS_REALTIME_RESOLUTION](https://github.com/yutiansut/QUANTAXIS_REALTIME_RESOLUTION) 实时交易/部署解决方案(未开源)
- [QUANTAXIS UNICORN](https://github.com/yutiansut/quantaxis_unicorn) QUANTAXIS 策略托管, 交易监控解决方案(未开源)
- [QUANTAXIS_RANK](https://github.com/yutiansut/QARank) QUANTAXIS实时账户评估
- [QUANTAXIS_CEPEngine](https://github.com/yutiansut/QACEPEngine) QUANTAXIS 复杂事件处理引擎
- [QUANTAXIS_PBSystem](https://github.com/yutiansut/QAPBSystem) QUANTAXIS PB系统
- [QUANTAXIS_QARISKPRO](https://github.com/yutiansut/QARISKPRO) QUANTAXIS 多市场多账户集成的实时风控系统
- [QUANTAXIS QADESKPRO](https://github.com/yutiansut/qadeskpro) 新版本客户端网页(部分开源)
- [QUANTAXIS PMS](https://github.com/yutiansut/QAPMS) 一个轻量级的纯python实现的  兼容QIFI协议的账户/仓位管理系统

> tick回测

- [QUANTAXIS TICKBacktest](https://github.com/yutiansut/QATickBacktest) tick回测 支持真实tick/仿真tick

> jupyterhub 定制(多人编辑)

- [QUANTAXIS JUPYTERHUB](https://github.com/yutiansut/QAJupyter)

> docker cluster

- [QUANTAXIS PROCluster](https://github.com/yutiansut/QAPRO_dockercluster) 一键部署的docker集群, 2地3中心的高可用灾备投研/交易环境


### 社区提供的工具链

- [QUANTAXIS_MONITOR_GUI](https://github.com/QUANTAXIS/QUANTAXIS_Monitor_GUI) 基于QT的python监控
- (目前废弃)[QUANTAXIS_DESKTOP](https://github.com/QUANTAXIS/QADESKTOP) 基于VUE.js/ ELECTRON的 桌面终端
- [portable_QA](https://github.com/QUANTAXIS/portable_QA) 一个独立的python环境,免配置
- [QUANTAXIS_CRAWLY](https://github.com/QUANTAXIS/QUANTAXIS_CRAWLY) 爬虫部分



==========================================


## 视频录播

### (2019-03-23)QUANTAXIS 2019 支持多市场的面向场景解决方案

[哔哩哔哩 传送门](https://www.bilibili.com/video/av47284727/)

### 2019-09-07 STU02 第二次录播 关于docker和流计算

[哔哩哔哩 传送门](https://www.bilibili.com/video/av67022443)

### 2019-09-14 STU03 第三次录播 关于期货模拟盘

[哔哩哔哩 传送门](https://www.bilibili.com/video/av67817623)


==========================================
<!-- TOC -->

- [QUANTAXIS 量化金融策略框架](#quantaxis-量化金融策略框架)
    - [0. 一键部署环境](#0-QASERVICE)
    - [1. 功能](#1-功能)
        - [1.1 行情服务](#11-行情服务)
            - [1.1.1 股票/期货/期权/美股/外汇/宏观的历史/实时行情(日线/分钟线/tick/实时五档)服务](#111-股票期货期权美股外汇宏观的历史实时行情日线分钟线tick实时五档服务)
            - [1.1.2 财务/基本面/宏观数据](#112-财务基本面宏观数据)
            - [1.1.3 自定义数据源的数据](#113-自定义数据源的数据)
        - [1.2 数据运维服务](#12-数据运维服务)
        - [1.3 分析服务](#13-分析服务)
            - [1.3.1 专门为A股股票数据适配的数据结构](#131-专门为a股股票数据适配的数据结构)
            - [1.3.2 精心为A股指标计算适配的指标类](#132-精心为a股指标计算适配的指标类)
        - [1.4 可扩展事件驱动框架](#14-可扩展事件驱动框架)
        - [1.5 回测服务](#15-回测服务)
            - [1.5.1 股票/日内t0/ 的日线/分钟线级别回测](#151-股票日内t0-的日线分钟线级别回测)
        - [1.6 实盘](#16-实盘)
            - [1.6.1 股票/实盘易 QAShiPaneBroker](#161-股票-实盘易--qashipanebroker)
            - [1.6.2 股票/TTS    QATTSBroker](#162-股票-tts----qattsbroker)
            - [1.6.3 期货/Open-trade-gateway QAOTGBROKER](#163-期货--qaotgbroker)
            - [1.6.4 期货/海风AT  QA_ATBroker](#164-期货--海风at封装-qaatbroker)
            - [1.6.5 期货/VNPY   QAVNPYBroker](#165-期货-vnpy-qavnpybroker)
        - [1.7 网站HTTP服务](#17-网站http服务)
            - [1.7.1 网站后台标准化接口](#171-网站后台标准化接口)
    - [2. 文档](#2-文档)
    - [3. 安装和部署](#3-安装和部署)
        - [3.0 安装说明](#30-安装说明)
        - [3.1 小白式上手](#31-小白式上手)
        - [3.2 部署式安装](#32-部署式安装)
        - [3.3  本地代码 开发式安装](#33--本地代码-开发式安装)
        - [3.4 代码提交式安装](#34-代码提交式安装)
    - [4. 更新](#4-更新)
    - [5. Docker](#5-docker)
    - [6. 使用说明](#6-使用说明)
    - [7. Jupyter示例](#7-jupyter示例)
    - [8. 开发计划](#8-开发计划)
    - [9. 常见问题FAQ](#9-常见问题faq)
    - [10. 项目捐赠](#10-项目捐赠)
    - [11. 回测Webkit插件概览](#11-回测webkit插件概览)
    - [12. QUANTAXIS 标准化协议和未来协议](#12-quantaxis-标准化协议和未来协议)
    - [13. 电脑配置推荐](#13-电脑配置推荐)
    - [14. 视频录播](#14-录播)
    - [License](#license)

<!-- /TOC -->




## 0. QASERVICE

qaservice是 quantaxis 面向(不想自己配环境的小白) / (有超级多服务器的网管大佬提供的) 一键配置环境

虽然qa-service一个命令就可以使用了 但是作者非常希望你可以读一下这个好不容易才写出来的文章

文章链接 [关于docker 和 QASERVICE](docker/readme.md)



##  1. 功能
======

![QUANTAXIS 2019.png](http://pic.yutiansut.com/FnRlMW2LQpFBrsdRv7E_uJ9RvzHt)

已经实现：

### 1.1 行情服务

#### 1.1.1 股票/期货/期权/美股/外汇/宏观的历史/实时行情(日线/分钟线/tick/实时五档)服务

参见 [QUANTAXIS的数据获取指南](Documents/DataFetch.md)

#### 1.1.2 财务/基本面/宏观数据

参见 [QUANTAXIS财务指标](Documents/financial_means.md)

#### 1.1.3 自定义数据源的数据

参见 [QUANTAXIS WEB 爬虫](Documents/crawler.md)

### 1.2 数据运维服务

一键更新 参见[WINDOWS数据自动更新](config/windows_autojob_updatedata.md)


### 1.3 分析服务

#### 1.3.1 专门为A股股票数据适配的数据结构

参见 [QUANTAXIS的数据结构](Documents/DataStruct.md)

参见 [QUANTAXIS行情研究](Documents/analysis.md)

#### 1.3.2 精心为A股指标计算适配的指标类

参见 [QUANTAXIS指标系统](Documents/indicators.md)


### 1.4 可扩展事件驱动框架

参见 [QUANTAXIS事件框架](Documents/about_event.md)


### 1.5 回测服务

#### 1.5.1 股票/日内t0/ 的日线/分钟线级别回测

参见 [QUANTAXIS的账户结构](Documents/account.md)

参见 [QUANTAXIS 账户风险分析插件指南](Documents/risk.md)

参见 [QUANTAXIS回测委托成交结算的说明](Documents/orderanddeal.md)

参见 [QUANTAXIS回测分析](Documents/backtestanalysis.md)

参见 [常见策略整理](Documents/strategy.md)

参见 [简单策略回测详解](Documents/QUANTAXIS回测分析全过程讲解.md)

参见 [T0交易的账户详解](Documents/TEST_ORDER_BACKTEST_T0.md)

### 1.6 实盘

#### 1.6.1 (股票) 实盘易  QAShipaneBroker(推荐)


实盘易插件 参见[实盘易](http://www.iguuu.com/e?x=18839)

实盘易安装注意 参见[安装注意](shipane_install_memo.md)

####  1.6.2 (股票) TTS    QATTSBroker

具体参见[QATTSBroker](QUANTAXIX/QAMARKET/QATTSBroker.py)

#### 1.6.3 (期货)  QAOtgBroker

CTP 的websocket交易 :  simnow demo:  ws://www.yutiansut.com:7988

关于websocket交易(可能存在一些不确定的bug 因此暂时不开放):

目前已经开源 参见 [QAOTGBROKER](https://github.com/yutiansut/QAOTGBROKER)

### 1.6.4 (期货) QATRADER(推荐)

CTP 的websocket 封装 包括完整的账户/HTTP控制等

目前已经开源 参见 [QATRADER](https://github.com/yutiansut/QATRADER)

#### 1.6.5 (期货)  海风AT封装 QAAtBroker

CTP 的海风broker魔改封装:  [QA_AtBroker](https://github.com/QUANTAXIS/QA_AtBroker)


#### 1.6.6 (期货) VNPY QAVNPYBroker

目前等VNPY 2.0的横空出世, 兼容py3.6/3.7后将进行对接



### 1.7 网站HTTP服务(目前已经独立为 QUANTAXIS_WEBSERVICE)

#### 1.7.1 网站后台标准化接口

参见 [QUANTAXIS WEB API说明](Documents/backendapi.md)




##  2. 文档

文档参见: [book](http://book.yutiansut.com)

下载文档手册(实时更新)

[PDF](https://www.gitbook.com/download/pdf/book/quantaxis/quantaxis) | [MOBI](https://www.gitbook.com/download/mobi/book/quantaxis/quantaxis) | [EPUB](https://www.gitbook.com/download/epub/book/quantaxis/quantaxis)

##  3. 安装和部署

```
quantaxis 1.4.4 增加了cython编译加速过程, 如果需要cython的同学 使用:

python setup_cython.py build_ext --inplace

再继续 pip install -e .

必须要注意的事情: 以上为可选项 并非强制!!


cython 提速:

1.4.3的Account.receive_simpledeal( 包括接受订单/更新账户) 为单次 1.88微秒
1.4.4的Account.receive_simpledeal  为单次 570纳秒
```

### 3.0 安装说明
参见 [安装说明](Documents/install.md)

### 3.1 小白式上手


参见 [小白上手教程WIN](Documents/install_for_new_to_coding.md)

### 3.2 部署式安装


```
pip install quantaxis -U
```
### 3.3  本地代码 开发式安装

本地安装
```
git clone https://github.com/quantaxis/quantaxis --depth 1

cd quantaxis

pip install -e .
```
### 3.4 代码提交式安装

代码提交式安装 代码提交参见  [代码提交](https://github.com/QUANTAXIS/QUANTAXIS/blob/master/Documents/about_pr.md)

- fork QUANTAXIS 到你的github账户

```
git clone https://github.com/你的账户名/quantaxis
```

##  4. 更新
参见 [更新说明](Documents/update.md)

##  5. Docker
参见 [Docker](Documents/docker.md)
##  6. 使用说明
参见 


* [QUANTAXIS的使用示例](https://github.com/quantaxis/QADemo)


##  7. Jupyter示例
参见 [Jupyter示例](jupyterexample)


##  8. 开发计划
参见 [开发计划](job_list.md)

##  9. 常见问题FAQ

参见 [FAQ](Documents/FAQ.md)

##  10. 项目捐赠

写代码不易...请作者喝杯咖啡呗?



CHANGELOG.md
# QUANTAXIS 更新纪要

<!-- TOC -->

- [QUANTAXIS 更新纪要](#quantaxis-更新纪要)
    - [1.3.0 ](#130)
    - [1.2.9 ](#129)
    - [1.2.8 ](#128)
    - [1.2.7 ](#127)
    - [1.2.6 ](#126)
    - [1.2.5 ](#125)
    - [1.2.4 ](#124)
    - [1.2.3 ](#123)
    - [1.2.2 ](#122)
    - [1.2.1 ](#121)
    - [1.2.0](#120)
    - [1.1.10](#1110)
    - [1.1.9](#119)
    - [1.1.8](#118)
    - [1.1.7](#117)
    - [1.1.6](#116)
    - [1.1.5](#115)
    - [1.1.4.dev1](#114dev1)
    - [1.1.4](#114)
    - [1.1.3.dev3](#113dev3)
    - [1.1.3.dev2](#113dev2)
    - [1.1.3.dev1](#113dev1)
    - [1.1.3](#113)
    - [1.1.2.dev1](#112dev1)
    - [1.1.2](#112)
    - [1.1.1.dev2](#111dev2)
    - [1.1.1.dev1](#111dev1)
    - [1.1.0](#110)
    - [1.0.68](#1068)
    - [1.0.67](#1067)
    - [1.0.66](#1066)
    - [1.0.65](#1065)
    - [1.0.64](#1064)
    - [1.0.63](#1063)
    - [1.0.62](#1062)
    - [1.0.61](#1061)
    - [1.0.60](#1060)
    - [1.0.59](#1059)
    - [1.0.58](#1058)
    - [1.0.57](#1057)
    - [1.0.56](#1056)
    - [1.0.55](#1055)
    - [1.0.54](#1054)
    - [1.0.53](#1053)
    - [1.0.52](#1052)
    - [1.0.51](#1051)
    - [1.0.50](#1050)
    - [1.0.49](#1049)
    - [1.0.48](#1048)
    - [1.0.47](#1047)
    - [1.0.46](#1046)
    - [1.0.45](#1045)
    - [1.0.44](#1044)
    - [1.0.43](#1043)
    - [1.0.42](#1042)
    - [1.0.41](#1041)
    - [1.0.40](#1040)
    - [1.0.39](#1039)
    - [1.0.38](#1038)
    - [1.0.37](#1037)
    - [1.0.36](#1036)
    - [1.0.35](#1035)
    - [1.0.34](#1034)
    - [1.0.33](#1033)
    - [1.0.32](#1032)
    - [1.0.31](#1031)
    - [1.0.30](#1030)
    - [1.0.29](#1029)
    - [1.0.28](#1028)
    - [1.0.27](#1027)
    - [1.0.26](#1026)
    - [1.0.25](#1025)

<!-- /TOC -->
## 1.3.0

1. QAUSER/ QAPORTFOLIO/ QAACCOUNT 的联动更新
2. 账户/组合/用户的刷新与恢复

```

In [2]: import QUANTAXIS as QA

In [3]: user1= QA.QA_User(username='yutiansut',password='940809')

In [4]: user1
Out[4]: < QA_USER USER_b87YCy3U with 1 portfolio: {'RB_PORTFOLIO': < QA_Portfolio RB_PORTFOLIO with 3 Accounts >} >

In [5]: user1.portfolio_list
Out[5]: {'RB_PORTFOLIO': < QA_Portfolio RB_PORTFOLIO with 3 Accounts >}

In [6]: user1['RB_PORTFOLIO']
Out[6]: < QA_Portfolio RB_PORTFOLIO with 3 Accounts >

In [7]: user1['RB_PORTFOLIO'].accounts
Out[7]:
{'test1': < QA_Account test1 market: stock_cn>,
 'test2': < QA_Account test2 market: stock_cn>,
 'test3_future': < QA_Account test3_future market: stock_cn>}

In [8]: user1['RB_PORTFOLIO']['test1']
Out[8]: < QA_Account test1 market: stock_cn>

```



## 1.2.9
1. QAIndicator部分指标修正,base.py增加IFAND函数,indicator.py部分指标修正

## 1.2.8

1. market_preset 兼容 tdx的主连/指数获取 如JL8, JL9, RBL8 等
2. base指标 增加 BARLAST
3. market_preset 增加基础函数支持 get_exchange, get_name等
4. QA_USER 完善, 增加订阅策略和积分系统
5. 删除QADataStruct_min类中的 high_limit/low_limit, 修复daystruct的nextdayhighlimit字段
6. 增加tusharepro部分数据到postgresql的储存和调取方式
7. QATTSBroker发布


## 1.2.7

迁移目录:

1. 拆分 QUANTAXIS_CRAWLY 至 https://github.com/QUANTAXIS/QUANTAXIS_CRAWLY

- 减少twisted安装问题
- 模块解耦 功能分离

2. 拆分 QUANTAXIS_MONITOR_GUI 至 https://github.com/QUANTAXIS/QUANTAXIS_Monitor_GUI

- gui部分以插件形式提供

3. 使用yapf 大量修正格式





## 1.2.6

1. 优化QADOCKER文档
2. 增加QAORDER文档
3. 优化了QALog模块在打印大量日志的时候无法知道其用途的问题

现在的quantaxis log 会以这个模式作为name:
```
 'quantaxis_{}-{}-.log'.format(get_config(), os.sep, os.path.basename(sys.argv[0]).split('.py')[0], str(datetime.datetime.now().strftime(
        '%Y-%m-%d-%H-%M-%S')))
```
4. 修改了QAUser的注册模块逻辑
5. 增加了 QA_DataStruct_Min 和 QA_DataStruct_Day两个基类模型
6. 修复settle的一个bug



## 1.2.5

1. 对于QA.QA_util_code_tostr 增加 原先为list类型的支持 现在支持自动补全的  int/list/str 类型转 str
2. QA_DataStruct增加bar_gen  返回迭代的dataframe
3. 增加对于期货的日结算支持:

在期货的保证金模型中:

开仓会冻结保证金到frozen里面, 这时钱并未参与结算/  需要先在每日结算时结转冻结的保证金
4. 增加了QA_Account的 history_table字段,增加了一个  frozen字段, 用于记录历史的保证金增加
5. 大幅删减QA_Account的 receive_deal 精简代码  直接调用receive_simpledeal方法
6. QADATASTRUCT 增加一个 kline_echarts 方法 直接返回 echarts.kline类 可以在jupyter notebook中直接显示
7. QALog模块的默认输出为warning级别, 减少别的模块的无聊输出(点名: 尤其是macropy 疯狂输出)

plot的图示例:

- 可能需要先升级pyecharts 到最新版本 (```pip install pyecharts -U -i https://pypi.doubanio.com/simple```)

![](http://pic.yutiansut.com/QQ%E5%9B%BE%E7%89%8720190103220819.png)


## 1.2.4

1. 修复保证金的bug

PS: 真丢人 年末写个bug 2333


## 1.2.3 

1. 感谢@追梦, QA_Account的receive_simpledeal的成交方式中的 股票市场的印花税计算修正
2. 改写@地下的地下铁, 修正QA_Risk中计算assets的一个停牌无数据的bug
3. 感谢@风筝 修复了单标的多市场的获取bug
4. 重大修改:  增加保证金账户的支持(期货)


具体示例  参见: https://github.com/QUANTAXIS/QUANTAXIS/blob/master/EXAMPLE/test_backtest/FUTURE/TEST_%E4%BF%9D%E8%AF%81%E9%87%91%E8%B4%A6%E6%88%B7.ipynb

```python
#在QA_Account的初始化的时候带上 allow_margin=True

acc=QA.QA_Account(allow_sellopen=True,init_cash=10000,allow_t0=True,allow_margin=True,account_cookie='future_test',market_type=QA.MARKET_TYPE.FUTURE_CN,frequence=QA.FREQUENCE.FIFTEEN_MIN)



#快速撮合接口的测试

acc.reset_assets(init_cash=10000)

acc.receive_simpledeal(code='RB1901', trade_price=3420, trade_amount=1, trade_towards=QA.ORDER_DIRECTION.BUY_OPEN, trade_time='2018-12-28 09:30:00')

acc.receive_simpledeal(code='RB1901', trade_price=3425, trade_amount=1, trade_towards=QA.ORDER_DIRECTION.SELL_CLOSE, trade_time='2018-12-28 09:45:00')

acc.receive_simpledeal(code='RB1901', trade_price=3435, trade_amount=1, trade_towards=QA.ORDER_DIRECTION.SELL_OPEN, trade_time='2018-12-28 09:55:00')

acc.receive_simpledeal(code='RB1901', trade_price=3420, trade_amount=1, trade_towards=QA.ORDER_DIRECTION.BUY_CLOSE, trade_time='2018-12-28 10:45:00')

acc.history_table
"""
datetime	code	price	amount	cash	order_id	realorder_id	trade_id	account_cookie	commission	tax	message
0	2018-12-28 09:30:00	RB1901	3420	1	6918.580	None	None	None	future_test	3.420	0	None
1	2018-12-28 09:45:00	RB1901	3425	-1	10038.155	None	None	None	future_test	3.425	0	None
2	2018-12-28 09:55:00	RB1901	3435	-1	6943.220	None	None	None	future_test	3.435	0	None

setup_cython.py
# coding=utf-8
#
# The MIT License (MIT)
#
# Copyright (c) 2016-2017 yutiansut/QUANTAXIS
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import codecs
import io
import os
import re
import sys
import webbrowser
import platform
import configparser
try:
    from setuptools import setup, Extension
except:
    from distutils.core import setup
"""
"""

if sys.version_info.major != 3 or sys.version_info.minor not in [4, 5, 6, 7, 8]:
    print('wrong version, should be 3.4/3.5/3.6/3.7/3.8 version')
    sys.exit()

with io.open('QUANTAXIS/__init__.py', 'rt', encoding='utf8') as f:
    context = f.read()
    VERSION = re.search(r'__version__ = \'(.*?)\'', context).group(1)
    AUTHOR = re.search(r'__author__ = \'(.*?)\'', context).group(1)


try:
    if sys.platform in ['win32', 'darwin']:
        print(webbrowser.open(
            'https://github.com/QUANTAXIS/QUANTAXIS/releases'))
        print('finish install')
except:
    pass


def read(fname):

    return codecs.open(os.path.join(os.path.dirname(__file__), fname)).read()


NAME = "quantaxis"
"""

"""
PACKAGES = ["QUANTAXIS", "QUANTAXIS.QAFetch", "QUANTAXIS.QACmd", "QUANTAXIS.QAMarket", 'QUANTAXIS.QASetting', "QUANTAXIS.QACmd",
            "QUANTAXIS.QAApplication", "QUANTAXIS.QAEngine", "QUANTAXIS.QAData", 'QUANTAXIS.QAData.proto', "QUANTAXIS.QAAnalysis", 'QUANTAXIS.QASelector',
            "QUANTAXIS.QASU", "QUANTAXIS.QAUtil", "QUANTAXIS.QAARP", "QUANTAXIS.QAIndicator"]
"""

"""

DESCRIPTION = "QUANTAXIS:Quantitative Financial Strategy Framework"

with open("README_ENG.md", "r") as fh:
    LONG_DESCRIPTION = fh.read()

"""

"""

KEYWORDS = ["quantaxis", "quant", "finance", "Backtest", 'Framework']
"""

"""

AUTHOR_EMAIL = "yutiansut@qq.com"

URL = "https://github.com/quantaxis/quantaxis"


LICENSE = "MIT"


ext_modules = [
    Extension('QUANTAXIS.QAARP.QAAccount', ['QUANTAXIS/QAARP/QAAccount.pyx']),
    Extension('QUANTAXIS.QAMarket.QAOrder', ['QUANTAXIS/QAMarket/QAOrder.pyx'])]

setup(
    name=NAME,
    version=VERSION,
    description=DESCRIPTION,
    long_description=LONG_DESCRIPTION,
    classifiers=[
        'License :: OSI Approved :: MIT License',
        'Programming Language :: Python',
        'Intended Audience :: Developers',
        'Operating System :: OS Independent',
    ],
    install_requires=['pandas>=0.23.4', 'numpy>=1.12.0', 'tushare', 'flask_socketio>=2.9.0 ', 'motor>=1.1', 'seaborn>=0.8.1', 'pyconvert>=0.6.3',
                      'lxml>=4.0', ' beautifulsoup4', 'matplotlib', 'requests', 'tornado', 'janus',
                      'demjson>=2.2.4', 'pymongo>=3.7', 'six>=1.10.0', 'tabulate>=0.7.7', 'pytdx>=1.67', 'retrying>=1.3.3',
                      'zenlog>=1.1', 'delegator.py>=0.0.12', 'flask>=0.12.2', 'pyecharts', 'protobuf>=3.4.0'],
    entry_points={
        'console_scripts': [
            'quantaxis=QUANTAXIS.QACmd:QA_cmd',
            'quantaxisq=QUANTAXIS.QAFetch.QATdx_adv:bat',
            'qarun=QUANTAXIS.QACmd.runner:run'
        ]
    },
    ext_modules=ext_modules,
    # install_requires=requirements,
    keywords=KEYWORDS,
    author=AUTHOR,
    author_email=AUTHOR_EMAIL,
    url=URL,
    license=LICENSE,
    packages=PACKAGES,
    include_package_data=True,
    zip_safe=True
)
